X11R6/xc/doc/man/Xi/XChgKbd.man
.\"
.\" Copyright ([\d,\s]*) by Hewlett-Packard Company, Ardent Computer, 
.\" 
.\" Permission to use, copy, modify, distribute, and sell this documentation 
.\" for any purpose and without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\" Ardent, and Hewlett-Packard make no representations about the 
.\" suitability for any purpose of the information in this document.  It is 
.\" provided \`\`as is'' without express or implied warranty.
.\" 
.\" $XConsortium: XChgKbd.man,v 1.5 94/06/04 17:32:20 rws Exp $
.ds xL Programming With Xlib
.TH XChangeKeyboardDevice 3X11 "Release 6" "X Version 11" "X FUNCTIONS"
.SH NAME
XChangeKeyboardDevice \- change which device is used as the X keyboard
.SH SYNTAX
Status XChangeKeyboardDevice\^(\^\fIdisplay\fP, \fIdevice\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      XDevice *\fIdevice\fP\^; 
.br
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I device
Specifies the device to be used as the X keyboard.
.SH DESCRIPTION
The \fIXChangeKeyboardDevice\fP request causes the server to use the specified
device as the X keyboard.  The device must have been previously opened by the
requesting client via \fIXOpenDevice\fP or a \fIBadDevice\fP error will result.
The device must support input class \fIKeys\fP, or a \fIBadMatch\fP error
will result.  If the server implementation does not support using the requested
device as the X keyboard, a \fIBadDevice\fP error will result.
.LP
If the specified device is grabbed by another client, \fIAlreadyGrabbed\fP
is returned.  If the specified device is frozen by a grab on another device,
\fIGrabFrozen\fP is returned.
If the request is successful, \fISuccess\fP is returned.
.LP
If the request succeeds,
a \fIChangeDeviceNotify\fP event is sent to all clients that have selected that
event.  A \fIMappingNotify\fP event with request = \fIMappingKeyboard\fP
is sent to all clients.
The specified device becomes the X keyboard and 
the old X keyboard becomes accessible through the input extension
protocol requests.
.LP
\fIXChangeKeyboardDevice\fP can generate a \fIBadDevice\fP or a \fIBadMatch\fP
error.
.SH DIAGNOSTICS
.TP 12
\fIBadDevice\fP
An invalid device was specified.  The specified device does not exist, has 
not been opened by this client via \fIXOpenInputDevice\fP,  or is already
one of the core X device (pointer or keyboard).  This error may
also occur if 
the server implementation does not support using the specified device as
the X keyboard.
.TP 12
\fIBadMatch\fP
This error may occur if an \fIXChangeKeyboardDevice\fP request was made 
specifying a device that has no keys.
.SH "SEE ALSO"
XChangePointerDevice
.br
\fI\*(xL\fP


X11R6/xc/lib/Xi/XChgKbd.c
/* $XConsortium: XChgKbd.c,v 1.8 94/04/17 20:17:57 rws Exp $ */

/************************************************************

Copyright (c) 1989  X Consortium

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of the X Consortium shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from the X Consortium.

Copyright (c) 1989 by Hewlett-Packard Company, Palo Alto, California.

			All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the name of Hewlett-Packard not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

HEWLETT-PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HEWLETT-PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

********************************************************/

/***********************************************************************
 *
 * XChangeKeyboardDevice - Change the device used as the X keyboard.
 *
 */

#include "XI.h"
#include "XIproto.h"
#include "Xlibint.h"
#include "XInput.h"
#include "extutil.h"

int
XChangeKeyboardDevice (dpy, dev)
    register Display 	*dpy;
    XDevice		*dev;
    {       
    xChangeKeyboardDeviceReq 	*req;
    xChangeKeyboardDeviceReply 	rep;
    XExtDisplayInfo *info = (XExtDisplayInfo *) XInput_find_display (dpy);

    LockDisplay (dpy);
    if (_XiCheckExtInit(dpy, XInput_Initial_Release) == -1)
	return (NoSuchExtension);

    GetReq(ChangeKeyboardDevice,req);		
    req->reqType = info->codes->major_opcode;
    req->ReqType = X_ChangeKeyboardDevice;
    req->deviceid = dev->device_id;
    rep.status = Success;

    (void) _XReply (dpy, (xReply *) &rep, 0, xTrue);

    UnlockDisplay(dpy);
    SyncHandle();
    return (rep.status);
    }

